### PR TITLE
Control Cluster Shards Balancing

### PR DESCRIPTION
**Pull request by hanbj**
Previous implementations were well thought out, but an index balance would cause all indexes to balance together.

### CODE REVIEW

**Review by ywelsch — COMMENTED** [review]
Thank you for your interest in contributing to ES. I don't understand what this change is trying to achieve, unfortunately. Can you provide some more explanations? In particular, why is it removing the code that allows index-level settings from overriding the cluster-level ones?

**Comment by hanbj** [comment]
If an index is allowed to be rebalance, then this line of code "if (allocation. deciders (). canRebalance (allocation). type ()!= Type. YES)" returns false, and the code continues to execute downward. In the balanceByWeights () method, all indexes are traversed according to the weight rebalance, so I set cluster. routing. rebalance. enable: "none" is invalid.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationDecider.java [hunk 1]
```diff
@@ -147,17 +147,29 @@ public Decision canRebalance(RoutingAllocation allocation) {
...
-                    return allocation.decision(Decision.YES, NAME, "rebalancing is permitted on one or more indices");
+        Rebalance enable = this.enableRebalance;
+        switch (enable) {
+            case ALL:
+                return allocation.decision(Decision.YES, NAME, "all rebalance are allowed");
+            case NONE:
+                return allocation.decision(Decision.NO, NAME, "none rebalance are not allowed");
+            case PRIMARIES:
+                if (allocation.routingNodes().hasInactivePrimaries()) {
+                    return allocation.decision(Decision.NO, NAME,
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationDecider.java]: why does this globally disable rebalancing when there are some inactive primaries of an unrelated index?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationDecider.java [hunk 2]
```diff
@@ -147,17 +147,29 @@ public Decision canRebalance(RoutingAllocation allocation) {
...
+                            "the cluster has inactive primary shards and cluster setting [%s] is set to [%s]",
+                            CLUSTER_ROUTING_REBALANCE_ENABLE_SETTING, enable);
                 }
-            }
-            return allocation.decision(Decision.NO, NAME, "no rebalancing is allowed due to %s", setting(enableRebalance, false));
+                return allocation.decision(Decision.YES, NAME, "all primary shards is active and rebalance are allowed");
+            case REPLICAS:
+                if (allocation.routingNodes().hasInactiveShards()) {
+                    return allocation.decision(Decision.NO, NAME,
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationDecider.java]: why does this globally disable rebalancing when there are inactive shards of some unrelated index?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationDecider.java [hunk 3]
```diff
@@ -147,17 +147,29 @@ public Decision canRebalance(RoutingAllocation allocation) {
...
-            for (IndexMetaData indexMetaData : allocation.metaData()) {
-                if (INDEX_ROUTING_REBALANCE_ENABLE_SETTING.exists(indexMetaData.getSettings())
-                    && INDEX_ROUTING_REBALANCE_ENABLE_SETTING.get(indexMetaData.getSettings()) != Rebalance.NONE) {
-                    return allocation.decision(Decision.YES, NAME, "rebalancing is permitted on one or more indices");
+        Rebalance enable = this.enableRebalance;
+        switch (enable) {
+            case ALL:
+                return allocation.decision(Decision.YES, NAME, "all rebalance are allowed");
+            case NONE:
+                return allocation.decision(Decision.NO, NAME, "none rebalance are not allowed");
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationDecider.java]: As said earlier, this changes the behavior of ES not to take the index-level property into account anymore when the cluster-level property is set.

**Review by ywelsch — CHANGES_REQUESTED** [review]
I'm not sure I understand what the purpose of this PR is, especially as it's disabling existing ES functionality. Is it to address a scaling issue? You mention a cluster with 300,000 shards on another PR, so I assume this is related. Why is this cluster making use of index-level rebalancing (i.e. the `index.routing.rebalance.enable` setting) if that is turning out to be a scaling bottleneck?

**Comment by ywelsch** [comment]
I don't see a way to take this PR forward (as it is breaking existing functionality), which is why I'm closing this. @hanbj do let me know about the problem this is trying to address and in particular whether it is something that would be covered by other work (e.g. https://github.com/elastic/elasticsearch/pull/42738)

**Comment by hanbj** [comment]
@ywelsch Thank you, there is indeed a problem at the moment, we are planning to implement it from another namespace way, ES cluster can be infinitely expanded, so this is the reason why I have not updated this pr for so long, sorry

**Pull request closed without merging**